### PR TITLE
Add relative and absolute residual norm stopping criteria

### PR DESCRIPTION
**Pull request by fritzgoebel**
This PR adds stopping criteria for relative (to norm(b)) and absolute residual norm.
It abstracts the three criteria `ResidualNormReduction,` `RelativeResidualNorm` and `AbsoluteResidualNorm,` which only differ in the way they initialize `starting_tau_` from `ResidualNorm`. 

If someone has a better name for the base class, I am open to suggestions, maybe something line `ResidualNormCriterion`?

Closes https://github.com/ginkgo-project/ginkgo/issues/147

### CODE REVIEW

File: core/stop/residual_norm.cpp [hunk 1]
```diff
@@ -30,53 +30,70 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
...
     }
     bool all_converged = true;
 
-    this->get_executor()->run(
-        residual_norm_reduction::make_residual_norm_reduction(
-            dense_tau, starting_tau_.get(), parameters_.reduction_factor,
-            stoppingId, setFinalized, stop_status, &this->device_storage_,
-            &all_converged, one_changed));
+    if (dense_tau->get_size()[1] > 1) {
+    }
```
Comment (yhmtsai) [inline core/stop/residual_norm.cpp:71]: Is anything missing here?

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace stop {
+
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/residual_norm.hpp:49]:
```suggestion


```

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    std::unique_ptr<NormVector> starting_tau_{};
+    std::unique_ptr<NormVector> u_dense_tau_{};
+
+private:
+    remove_complex<ValueType> tolerance_{};
+    /* Contains device side: all_converged and one_changed booleans */
+    Array<bool> device_storage_;
+};
+
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/residual_norm.hpp:84]:
```suggestion


```

File: include/ginkgo/core/stop/residual_norm.hpp (hunk 2 above)
Comment (yhmtsai) [reply include/ginkgo/core/stop/residual_norm.hpp:49]: Could you add some description of ResidualNorm?

**Comment by pratikvn** [comment]
I would still keep the abstract class name as `ResidualNormReduction` as we are still reducing the residual norm (but different types of the residual norm). The derived classes can be `AbsoluteResNormReduction` and `RelativeResNormReduction` or `AbsResNormReduction` and `RelResNormReduction` if the first two names are too long.

**Comment by upsj** [comment]
@pratik I would disagree there, since to me the Reduction in the name refers to the amount we reduce the residual norm by, while e.g. for Absolute/RelativeResidualNorm we give an exact tolerance, not a reduction factor.

File: omp/test/stop/residual_norm_kernels.cpp [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+template <typename T>
+class AbsoluteResidualNorm : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Dense<T>;
+    using NormVector = gko::matrix::Dense<gko::remove_complex<T>>;
+
+    AbsoluteResidualNorm()
+    {
+        std::cout << "TOLERANCE: " << r<T>::value << "\n";
```
Comment (pratikvn) [inline omp/test/stop/residual_norm_kernels.cpp:251]: Can be removed ?

**Comment by fritzgoebel** [comment]
> LGTM! Do you actually test with different values for the different tests? It didn't look to me like that was the case when skimming over the CUDA tests.

You mean different values for the tolerance? No, I use the same value as was used before for the ResidualNormReduction criterion..

File: include/ginkgo/core/stop/residual_norm_reduction.hpp [hunk 5]
```diff
@@ -30,100 +30,5 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
...
-    Array<bool> device_storage_;
-};
-
-
-}  // namespace stop
-}  // namespace gko
-
-
-#endif  // GKO_CORE_STOP_RESIDUAL_NORM_REDUCTION_HPP_
+// This file is deprecated and will be removed in a later release.
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/residual_norm_reduction.hpp:33]: Maybe try `#warning this file ....` as @tcojean said
I prefer use `in a later major release` or `in 2.0.0 release`

File: cuda/test/stop/residual_norm_kernels.cpp [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+            .residual_norm(d_res.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+    stop_status.set_executor(ref_);
+    ASSERT_TRUE(stop_status.get_data()[0].has_converged());
+    stop_status.set_executor(cuda_);
+    ASSERT_TRUE(one_changed);
+
+    res->at(0, 1) = tol * 1.0e+1;
+    d_res->copy_from(res.get());
+    ASSERT_TRUE(
```
Comment (yhmtsai) [inline cuda/test/stop/residual_norm_kernels.cpp:250]: In the previous test, WaitsTillResidualGoal, `res` needs to be `tol` to satisfy the stopping criterion, but `res` in this test only needs to `tol * 10`. Do I misunderstand somthing?

File: cuda/test/stop/residual_norm_kernels.cpp [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+    ASSERT_FALSE(
+        criterion->update()
+            .residual_norm(d_res.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+    stop_status.set_executor(ref_);
+    ASSERT_TRUE(stop_status.get_data()[0].has_converged());
+    stop_status.set_executor(cuda_);
+    ASSERT_TRUE(one_changed);
+
+    res->at(0, 1) = tol * 1.0e+1;
```
Comment (yhmtsai) [inline cuda/test/stop/residual_norm_kernels.cpp:354]: Shouldn't it converge?

File: cuda/test/stop/residual_norm_kernels.cpp (hunk 6 above)
Comment (fritzgoebel) [reply cuda/test/stop/residual_norm_kernels.cpp:252]: I got these wrong before, my idea was to start with the same initial residual and right-hand side for all criteria and then check the different behavior of the criteria:
- ResidualReduction is satisfied when `|r| / |r_0| < tol`, so when `|r| < 100 * tol`
- RelativeResidualNorm is staisfied when `|r| / |b| < tol`, so when `|r| < 10 * tol`
- AbsoluteResidualNorm is satisfied when `|r| < tol`

File: include/ginkgo/core/stop/residual_norm_reduction.hpp [hunk 8]
```diff
@@ -30,100 +30,10 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#ifndef GKO_CORE_STOP_RESIDUAL_NORM_REDUCTION_HPP_
-#define GKO_CORE_STOP_RESIDUAL_NORM_REDUCTION_HPP_
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/residual_norm_reduction.hpp:34]: Keep this?

File: test_install/test_install.cpp [hunk 9]
```diff
@@ -351,7 +351,7 @@ int main(int, char **)
         auto time = gko::stop::Time::build()
                         .with_time_limit(std::chrono::milliseconds(10))
                         .on(refExec);
-        // residual_norm_reduction.hpp
+        // residual_norm.hpp
         auto res_red = gko::stop::ResidualNormReduction<>::build()
```
Comment (tcojean) [inline test_install/test_install.cpp:355]: You need to add similar checks with the new RelativeResidualNorm and Absolute.

File: test_install/test_install_cuda.cu [hunk 10]
```diff
@@ -343,7 +343,7 @@ int main(int, char **)
         auto time = gko::stop::Time::build()
                         .with_time_limit(std::chrono::milliseconds(10))
                         .on(cudaExec);
-        // residual_norm_reduction.hpp
+        // residual_norm.hpp
```
Comment (tcojean) [inline test_install/test_install_cuda.cu:346]: Same here.

**Comment by tcojean** [comment]
Looks like the residual_norm_reduction.hpp file is still included in the main ginkgo.hpp although it should be removed. I think this file is generated by a script so the script needs to be updated to not include this file.
https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/616917734

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        this->u_dense_tau_ =
+            NormVector::create_with_config_of(this->starting_tau_.get());
+        dense_rhs->compute_norm2(this->starting_tau_.get());
+    }
+};
+
+
+/**
+ * The AbsoluteResidualNorm class is a stopping criterion which stops the
+ * iteration process when the residual norm is below a certain
```
[the remark this answers is not in the document]
Comment (fritzgoebel) [reply include/ginkgo/core/stop/residual_norm.hpp:217]: I think this should be clear with the equation and the difference in residual norm / residual norm relative to the initial residual / residual norm relative to the right-hand side in the description.

**Pull request merged**